table/block.cc

```diff
@@ -311,6 +348,42 @@ bool BlockIter::ParseNextKey() {
   }
 }
 
+bool IndexBlockIter::ParseNextIndexKey() {
+  current_ = NextEntryOffset();
+  const char* p = data_ + current_;
+  const char* limit = data_ + restarts_;  // Restarts come right after data
+  if (p >= limit) {
+    // No more entries to return.  Mark as invalid.
+    current_ = restarts_;
+    restart_index_ = num_restarts_;
+    return false;
+  }
+
+  // Decode next entry
+  uint32_t shared, non_shared, value_length;
+  p = DecodeEntry(p, limit, &shared, &non_shared, &value_length);
+  if (p == nullptr || key_.Size() < shared) {
+    CorruptionError();
+    return false;
+  }
+  if (shared == 0) {
+    // If this key dont share any bytes with prev key then we dont need
+    // to decode it and can use it's address in the block directly.
+    key_.SetKey(Slice(p, non_shared), false /* copy */);
+    key_pinned_ = true;
+  } else {
+    // This key share `shared` bytes with prev key, we need to decode it
+    key_.TrimAppend(shared, p, non_shared);
+    key_pinned_ = false;
+  }
+  value_ = Slice(p + non_shared, value_length);
+  while (restart_index_ + 1 < num_restarts_ &&
+         GetRestartPoint(restart_index_ + 1) < current_) {
+    ++restart_index_;
+  }
+  return true;
+}
+
 // Binary search in restart array to find the first restart point that
```

Comment:
@siying in the upcoming value delta encoding change DecodeEntry which is used by BlockIter::BinarySeek would be different in index blocks as it should decode the values differently if they are value delta encoded. However if make DecodeEntry a pointer function, it would disable inlining. We either need to complicate BlockIter by adding a variable indicating whether the value is delta encoded, or duplicate ::BinarySeek for index block. Which do you recommend?